### ISSUE TITLE
Incorrect metrics used on some fonts

### ISSUE THREAD

**Issue by rcombs** [libass/libass#361]
Sample: https://puu.sh/EKGA3/7130de7b82.ass
```
[Script Info]
ScriptType: v4.00+
PlayResX: 1920
PlayResY: 1080
Timer: 0.0000
WrapStyle: 0
YCbCr Matrix: TV.709
ScaledBorderAndShadow: yes

[Aegisub Project Garbage]
Export Encoding: Unicode (UTF-8)
Video File: ?dummy:24.000000:2700:1920:1080:255:254:254:c

[V4+ Styles]
Format: Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, OutlineColour, BackColour, Bold, Italic, Underline, StrikeOut, ScaleX, ScaleY, Spacing, Angle, BorderStyle, Outline, Shadow, Alignment, MarginL, MarginR, MarginV, Encoding
Style: Default,Profile Medium,70,&H00FFFFFF,&HC0000000,&H00000000,&HA8000000,0,0,0,0,95,100,0,0,1,3,1.5,2,150,150,60,1

[Events]
Format: Layer, Start, End, Style, Name, MarginL, MarginR, MarginV, Effect, Text
Dialogue: 10,0:00:00.00,0:00:06.35,Default,,0,0,0,,They showed clear animosity toward us, \Nunlike any other demonic beasts we've encountered before.
```

This renders as 3 lines in libass, at a larger font size than in VSFilter or _apparently_ older libass versions, though I've been unable to find a version on which the issue doesn't repro in order to bisect. I could be missing something obvious, though; it's late and I'm sleepy.

Looks to be a separate issue from #238?

**Comment by astiob** [comment]
Sounds the same as #352 though, which is currently kind of blocking a release. If anyone could bisect and find the bad commit, that would be great, but at some point I’ll try to find time to do it myself.

**Comment by rcombs** [comment]
Hmm, I just tried harfbuzz 1.0.3 and could still repro both this issue and #352 there. Going to try an older freetype now.

**Comment by rcombs** [comment]
This is a regression(?) in this FreeType commit: https://git.savannah.gnu.org/cgit/freetype/freetype2.git/commit/?id=f72b00746c64e13625cf8371f031411fbd0d6161
ML discussion: https://marc.info/?t=154508255300008&r=1&w=2

So it seems like FreeType has started obeying a metrics group (the OS/2 typo fields) that aren't used in GDI (which apparently prefers the horizontal table?).

This can be worked around using this patch:
```
diff --git a/libass/ass_font.c b/libass/ass_font.c
index d17b128..ea02866 100644
--- a/libass/ass_font.c
+++ b/libass/ass_font.c
@@ -97,6 +97,15 @@ uint32_t ass_font_index_magic(FT_Face face, uint32_t symbol)
 
 static void buggy_font_workaround(FT_Face face)
 {
+    // Mimicking GDI's behavior for asc/desc/height
+    TT_HoriHeader *hh = FT_Get_Sfnt_Table(face, ft_sfnt_hhea);
+    if (hh) {
+        face->ascender  = hh->Ascender;
+        face->descender = hh->Descender;
+        face->height = face->ascender - face->descender +
+                       hh->Line_Gap;
+    }
+
     // Some fonts have zero Ascender/Descender fields in 'hhea' table.
     // In this case, get the information from 'os2' table or, as
     // a last resort, from face.bbox.

```
This more or less restores the old FreeType behavior. However, rather than committing this as-is, I'd prefer to get some more information on what GDI does here. We have 3 sets of values we can consider (hhea, os2, and the bounding box). Can anyone edit a font to see what VSFilter does in various possible situations when those values all disagree, or when some are set to 0, or when some are missing altogether?

**Comment by downthomas** [comment]
If you're looking for more information on what GDI says to do, look at the OpenType specs:
[https://docs.microsoft.com/en-us/typography/opentype/spec/os2#uswinascent](https://docs.microsoft.com/en-us/typography/opentype/spec/os2#uswinascent)

I played around on my Windows 10 machine with a copy of [Profile Medium I got from fontsgeek](http://fontsgeek.com/fonts/Profile-Medium). Since it had different values for usWinAscent/usWinDescent, sTypoAscender/sTypoDescender, and hhea ascent/descent, I thought it would be an easy test. Funny enough, it displayed the same on MPC-HD/VSFilter, mpv, and VLC.

I was unable to reproduce this issue until I turned USE_TYPO_METRICS on in the font. Only then did mpv render the text into three lines. MPC-HD and VLC displayed the same as before. When I changed the sTypoAscender/sTypoDescender to match usWinAscent/usWinDescent with USE_TYPO_METRICS turned on, mpv rendered it on two lines again.

No changes to the hhea ascent/descent changed the render on any players. Zeroing out hhea ascent/descent made no difference in how the text was rendered.

Looking at the source code for VSFilter, it looks like they're using GetTextMetrics to get ascent/descent information, which supplies usWinAscent/usWinDescent. GetOutlineTextMetricsA gives access to sTypoAscender/sTypoDescender with otmMacAscent/otmMacDescent, but it doesn't look like that is what is used in practice.

Ultimately, I think it comes down to what standards you're going to follow and violate. Apple in its TrueType documentation says to use the hhea... unless you're on Windows because that uses the OS/2 table. Microsoft says that "The ascender, descender and linegap values in this [hhea] table are Apple specific." 

Freetype and GDI agree on the first step:
1. If the OS/2 table exists and the fsSelection bit 7 is set, (USE_TYPO_METRICS), trust the font and use the sTypo* metrics.

If the USE_TYPO_METRICS bit is set on a font, I think it is fair to use that. That is the font designer telling us, "Yes, really, really use the Typo metrics." [Microsoft strongly recommends using sTypo metrics when fsSelection bit 7 is set.](https://docs.microsoft.com/en-us/typography/opentype/spec/os2#fss) "If set, it is strongly recommended that applications use OS/2.sTypoAscender - OS/2.sTypoDescender + OS/2.sTypoLineGap as the default line spacing for this font."

If the font is designed to be cross-platform, it should display similarly regardless of whether the hhea table or the OS/2 table sTypo metrics are used.

**Comment by rcombs** [comment]
Thanks a lot for the analysis, it's really helpful!

For some background, in libass, we attempt to mimic VSFilter's behavior on existing scripts to the greatest extent possible; for all intents and purposes, VSFilter is considered the reference implementation, and defines the "standard" for how ASS is rendered. When we run into cases where different behavior seems more desirable, we sometimes add opt-in header flags (see e.g. ScaledBorderAndShadow), which are sometimes also implemented by the modern VSFilter forks.

So, it sounds like VSFilter's behavior is "Use usWinAscent/usWinDescent, regardless of what else the font says", right? What happens if those values are zero (or if the OS/2 table isn't present at all)? libass has some interesting behavior here; it uses usWinAscent/usWinDescent in `ass_face_set_size` and `ass_font_get_asc_desc` preferentially over the hhea table, but uses `face->ascender` and co in some other places… not sure how much of that is deliberate.

I think that if we want to define a new opt-in flag for more "correct" (i.e. consistent with expectations from other software) behavior, we should probably mimic FreeType's behavior pretty closely (perhaps just use it as-is?), but we might want to take a look at what other software does and see if anything else seems preferable. Regardless, we should define it clearly in our own docs, rather than just saying "if this flag is on, we do what FreeType does".

Also, while we're on the topic: you seem pretty knowledgeable about font internals; would you mind trying out some weight- and slant-based selection on VSFilter? (e.g. matching "{\b600\fnOpen Sans\i1}" or "{\fnOpen Sans Semibold\i1}" to Open Sans Semibold Italic). I'm wondering if there might also be multiple weight fields to consider, and if our current behavior is "correct" if they're inconsistent.

**Comment by downthomas** [comment]
Sorry for the delay in response, I had to think about this a little (ok more than a little) and I'm not quite done thinking.

Answers first:
1. VSFilter used usWinAscent/usWinDescent because it used GetTextMetrics.
2. If usWinAscent/usWinDescent is zeroed out in a font, Windows does not recognize it as a valid font.
3. If a font lacks an OS/2 table, Windows does not recognize it as a valid font, as the OS/2 table is mandatory on Windows. There may be some exceptions for bitmapped fonts, but I'm not sure.
4. There aren't multiple weight class fields to consider, and the current behavior of libass is correct. The weight and slant based selection behaves the same on VSFilter and libass.

Other thoughts:
1. From what I see in the libass code, usWinAscent/usWinDescent is used preferentially over the hhea table in ass_font_get_asc_desc, which is used by ass_outline_construct in ass_render.c. 
2. Otherwise, in ass_face_set_size, usWinAscent/usWinDescent is used to scale the ascender/descender hhea numbers. 
3. Otherwise, because libass used FT's ascender/descender, that preferred the hhea ascent/descent metrics.

Now that shouldn't matter in the first place. Generally, the hhea ascent/descent for a font should be the same as usWinAscent/usWinDescent, otherwise there will be different line heights on Windows and MacOS. They largely describe the same thing, the distance from baseline of the highest ascender/lowest descender. But some fonts are broken, some fonts don't have the right data, etc.

If the font is well designed, you can use the hhea metrics (ascent, descent and lineGap) to calculate line height. If the font is well designed, OS/2.sTypoAscender - OS/2.sTypoDescender + OS/2.sTypoLineGap should be used to calculate the baseline distance.

Part of the problem right now is that libass has been emulating but not following VSFilter, which was doing the wrong thing in using usWin*. Instead of doing what VSFilter was doing, which was wrong, it did a right thing but made it look like VSFilter.

Before the freetype change, ascender/descender came from the hhea table.
If the font had an OS/2 table, ass_face_set_size used ascender - descender from hhea and set that as ft_height.
It then set os2_height as usWinAscent + usWinDescent.
It sets Mscale = ft_height / os2_height and divides ascender/descender/height by it.
If the hhea height is the same as the usWinAscent/usWinDescent height, mscale is 1 and there's no change to the metrics. But if there is a difference, the font is scaled to look the same as if it was using usWinAscent/usWinDescent.

With the freetype change, for fonts that have fsSelection bit 7 set, ascender/descender comes from the OS/2 table.
Unless the sTypo* metrics are the same as the usWin* metrics (and they shouldn't be if fsSelection bit 7 is set), we will see a change in behavior.
If the font has an OS/2 table, ass_face_set_size uses ascender - descender from hhea and sets that as ft_height.
It then set os2_height as usWinAscent + usWinDescent.
It sets Mscale = ft_height / os2_height and divides ascender/descender/height by it.
But this time, ascender/descender/height is not coming from the hhea table, and the font isn't being scaled to the usWin* metrics height. It is applying the scale difference between hhea and usWin metrics to the OS/2 sTypo metrics. Most fonts with the bit set are going to be blown up in size unnecessarily.

There are a number of things that could be done, like some sort of option flag, but I also think that for fonts with fsSelection bit 7 set, simply not scaling ascender/descender might fix this behavior. I haven't had time to implement that to test it out.

I'm sure there are reasons why libass didn't following VSFilter in using usWin*, like it being wrong to use it in that way, but that's also an option to consider.

I meant to try to find out what others had done on this font issue, but I only stumbled upon how LibreOffice deals with it.

From what I understand, [LibreOffice uses](https://github.com/LibreOffice/core/commit/34d7602954d4483b3bc9db700e7df2c15348947a)
1. If USE_TYPO_METRICS flag is set, use sTypo metrics.
2. Otherwise use Win metrics if available.
2. Then fall back to hhea table.

Whereas Freetype's behavior now is:
1. If USE_TYPO_METRICS flag is set, use sTypo metrics.
2. Otherwise use hhea table.
3. Then use sTypo metrics.
4. Then use Win metrics.

Freetype's old behavior:
1. Use hhea table.
2. Then use sTypo metrics.
3. Then use Win metrics.

I'll try to implement at least one of the solutions I mentioned and come back with what I think.

**Comment by rcombs** [comment]
So, it sounds like the optimal compatibility route is:
1. If usWinAscent/usWinDescent are present and non-zero, use those (matching VSFilter for all fonts where it works at all)
2. Otherwise, if USE_TYPO_METRICS flag is set, use sTypo
3. Otherwise, use hhea
4. Otherwise, use sTypo
5. Otherwise, use the bbox

The `ass_face_set_size` routine was originally added in 676b104211c85790802ad38cdb6cc310d05b0484… and as far as I can tell, it shouldn't be needed at all if we switch `buggy_font_workaround` to unconditionally use `usWinAscent` + `usWinDescent` if available? (And then we'd probably want to rename it to something like `calculate_font_metrics`)

**Comment by downthomas** [comment]
Trying to determine the extent of the problem, I edited [ftdump](https://sourceforge.net/projects/freetype/files/freetype-demos/2.10.1/) to give me the various ascender/descender metrics in addition to the fsSelection bits. I ran it on all of my fonts, which came out to 4,672 font faces. Only 326 had `USE_TYPO_METRICS` set, 103 were TrueType files, 223 were OTF files, and the oldest goes back to 2007.

The one thing they all had in common: the information from the `hhea` table was the same in the sTypo fields. So `ascent = sTypoAscender,` `descent = sTypoDescender`, `line_gap = sTypoLineGap`. The same is true of the test font in #352 .

In other words, the change in FreeType can't be the cause of this regression because the numbers it was providing before the change and after the change have remained the same.

`ass_face_set_size` is used in [libass/ass_render.c](https://github.com/libass/libass/blob/0527f023f764976a6848b54af542c1141ae3f09e/libass/ass_render.c#L1168) and [libass/ass_shaper.c](https://github.com/libass/libass/blob/0527f023f764976a6848b54af542c1141ae3f09e/libass/ass_shaper.c#L448). It would be better to integrate buggy_font_workaround into `ass_face_set_size`.

**Comment by downthomas** [comment]
I've created some commits to use the sTypo* metrics for scaling in the appropriate cases.
2d4a4cf08dde5a4a46e69ae28ca9a667c05b9456 -- main edit 17e893adca4f7fbf286ca72ac60c534569dde932 -- with fixed indentation.

I tested it out and it appears to resolve the issue.

**Comment by downthomas** [comment]
More broadly on the optimal compatibility route:
1. The current route relies on freetype, which takes care of most of the ascender/descender issues. The one area it doesn't is with a buggy font with zero values in the hhea ascender/descender (basically, broken Windows fonts). buggy_font_workaround could be integrated into add_face, but `ass_face_set_size` should also deal with the case of when `ft_height = 0` and switch to using sTypo height.
2. As a practical matter, if we used usWinAscent/usWinDescent, we would never use sTypo. Any font with with a valid OS/2 table will have usWin* present and non-zero, so effectively it would be usWin -> hhea -> bbox.
3. Other than matching VSFilter output, I'm not sure there's any gain to be had. Libass already uses usWinAscent/usWinDescent for computing maximal ascender/descender. 
4. Typographically, using usWinAscent/usWinDescent for line height is incorrect, and libass should be using factoring in sTypoLineGap or hhea->line_gap.
5. Some function like ass_face_set_size will be necessary to calculate (or recalculate) font metrics.

**Comment by downthomas** [comment]
This commit would break any font that lacks an OS/2 table, which in general would be older Mac fonts and older TrueType fonts, and specifically, the fallback font Last Resort on macOS.
